File: mlpy/ModelDriver.py
from xml.dom.minidom import parseString, Node
from Models import *

class ModelDriver():
    TOOL_TAGS = ['Labels', 'Choices', 'BrushLabels']

    def __init__(self):
        pass

    @staticmethod
    # Parse the config of a project to get information(i.e. fromName, toName and type)
    def parseConfig(configs):
        # Denotes the finish of parsing config
        parseDone = False
        fromName = ''
        toName = ''
        toolType = ''

        # Retrieves nodes rooted at a given node recursively to extract information(i.e. fromName, toName and type)
        def extractInfo(parentNode):
            nonlocal parseDone, fromName, toName, toolType

            childNodes = parentNode.childNodes

            for childNode in childNodes:
                if (childNode.nodeType == Node.ELEMENT_NODE):
                    if (childNode.tagName in ModelDriver.TOOL_TAGS):
                        fromName = childNode.getAttribute("name")
                        toName = childNode.getAttribute("toName")
                        toolType = str(childNode.tagName).lower()
                        parseDone = True
                        break

            if(parseDone == False):
                if(len(childNodes) != 0):
                    for childNode in childNodes:
                        extractInfo(childNode)
            else:
                parseDone = False


        # Construct dom tree from config
        DOMTree = parseString(configs)

        # Get the root element of the dom tree
        rootElement = DOMTree.documentElement

        extractInfo(rootElement)

        return fromName, toName, toolType


    @staticmethod
    # Run model on a single data and update its predictions
    def runModelOnData(type, data):
        # TODO: Get configs from database
        configs = """<View>
                                <Image name="image" value="$image" zoom="true"/>
                                <BrushLabels name="tag" toName="image">
                                    <Label value="Airplane" background="rgba(255, 0, 0, 0.7)"/>
                                    <Label value="Car" background="rgba(0, 0, 255, 0.7)"/>
                                </BrushLabels>
                            </View>"""

        # TODO: Get pre-existing predictions from database
        predictions = []

        # TODO: Get modelVersion from database
        modelPath = '../ml/models/fcn/fcn.pth'

        # TODO: Get modelVersion from database
        modelVersion = 'undefined'

        fromName, toName, toolType = ModelDriver.parseConfig(configs)

        if(type == 'Semantic Segmentation Mask'):
            # TODO: Get index2nameMap from database
            index2nameMap = {
                0: 'Background',
                1: 'Aeroplane',
                2: 'Bicycle',
                3: 'Bird',
                4: 'Boat',
                5: 'Bottle',
                6: 'Bus',
                7: 'Car',
                8: 'Cat',
                9: 'Chair',
                10: 'Cow',
                11: 'Dining table',
                12: 'Dog',
                13: 'Horse',
                14: 'Motorbike',
                15: 'Person',
                16: 'Potted plant',
                17: 'Sheep',
                18: 'Sofa',
                19: 'Train',
                20: 'Tv/monitor'
            }

            model = SemSegMaskModel(modelPath, modelVersion, fromName, toName, type, index2nameMap)
        else:
            pass

        predictionItem = model.predict(data)
        # TODO: Merge predictionItems with the same model version together
        predictions.append(predictionItem)

        print(predictions)

        # TODO: Save new predictions in database


    @staticmethod
    # Run model on a list of data
    def runModelOnDatas(type, datas):
        for data in datas:
            ModelDriver.runModelOnData(type, data)


if __name__ == '__main__':
    ModelDriver.runModelOnData('Semantic Segmentation Mask', './puppy.webp')

File: consumer-server/src/main/java/com/grp202116/backend/ml/ModelDriver.java
package com.grp202116.backend.ml;

import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;
import com.grp202116.backend.mapper.PredictionMapper;
import com.grp202116.backend.pojo.ModelDO;
import com.grp202116.backend.pojo.PredictionDO;
import com.grp202116.backend.util.HttpUtils;
import org.springframework.beans.factory.annotation.Autowired;

import javax.annotation.Resource;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Objects;

//Todo: combine with the frontend to retrieve requested project id and data id;
public class ModelDriver {

    ModelDO model;

    public ModelDriver(BigInteger projectId) {
        this.model = new ModelDO();
    }

    public JSONObject parseConfig() {
        String configs = "<View>" +
                "  <Image name=\" image \" value=\" $image \" zoom=\" true \"/>" +
                " <BrushLabels name=\" tag \" toName=\" image \">" +
                "     <Label value=\" Airplane \" background=\" rgba(255, 0, 0, 0.7) \"/>" +
                "                               <Label value=\" Car \" background=\" rgba(0, 0, 255, 0.7) \"/>" +
                "                           </BrushLabels>" +
                "                       </View>";

        String labels = "Background, Aeroplane, Bicycle, Bird, Boat, Bottle, Bus, Car, Cat, Chair, Cow," +
                "Dining table, Dog, Horse, Motorbike, Person, Potted plant, Sheep, Sofa, Train," +
                " Tv/monitor";

        JSONObject param = new JSONObject();

        param.put("project_type", "Semantic Segmentation Mask");
        param.put("data", "./puppy.webp");
        param.put("configs", configs);
        param.put("model_path", "../ml/models/fcn/fcn.pth");
        param.put("model_version", "undefined");
        param.put("model_root", "./");
        param.put("labels", labels);

        return param;
    }

    public List<PredictionDO> savePredictions(JSONArray predictions) {
        List<PredictionDO> predictionList = new ArrayList<>();
        for (int i = 0; i < Objects.requireNonNull(predictions).size(); i++) {
            JSONObject predictionJSONObject = predictions.getJSONObject(i);
            PredictionDO prediction = new PredictionDO();
            prediction.setPredictionId(predictionJSONObject.getString("id"));
            prediction.setType(predictionJSONObject.getString("type"));
            prediction.setValue(predictionJSONObject.getString("value"));
            prediction.setFromName(predictionJSONObject.getString("from_name"));
            prediction.setToName(predictionJSONObject.getString("to_name"));

            Date date = new Date();
            prediction.setCreateTime(date);
            prediction.setUpdateTime(date);
            prediction.setDataId(new BigInteger(String.valueOf(1)));
            prediction.setProjectId(new BigInteger(String.valueOf(1)));

            predictionList.add(prediction);
        }

        return predictionList;
    }
}
